### PR TITLE
[v2.7] Disable IMR context save for MTL 

### PR DESCRIPTION
**Pull request by mengdonglin**
This is a follow up of PR #7994. We want to make sure it's safe to disable IMR context save for MTL and other platforms in sof v2.7

Please find more background info in https://github.com/thesofproject/sof/pull/7994#issuecomment-1706131833

We also want to check if disabling IMR context save can avoid the FW boot failure:
- https://github.com/thesofproject/sof/issues/7866#issuecomment-1700880432
- https://github.com/thesofproject/sof/issues/8148

### CODE REVIEW

**Review by tmleman — APPROVED** [review]

File: zephyr/lib/cpu.c [hunk 1]
```diff
@@ -148,7 +148,7 @@ int cpu_enable_core(int id)
 	 * initialization. By reinitializing the idle thread, we would overwrite the kernel structs
 	 * and the idle thread stack.
 	 */
-	if (!IS_ENABLED(CONFIG_ADSP_IMR_CONTEXT_SAVE) ||
+	if (!defined(CONFIG_ACE) ||
```
Comment (ranj063) [inline zephyr/lib/cpu.c:151]: @mengdonglin looks good. Can we also update the comment above to call out the difference between cavs and ACE?

Comment (mengdonglin) [reply zephyr/lib/cpu.c]: @ranj063 Thanks for your review. I updated the comments in the code.

**Review by RanderWang — APPROVED** [review]
It is a temporary solution for v2.7 and it is not fit to check hw platform in common framework. I can add dsp restore support on CAVS if I am free

File: zephyr/lib/cpu.c [hunk 2]
```diff
@@ -143,14 +143,25 @@ int cpu_enable_core(int id)
 		return 0;
 
 #if ZEPHYR_VERSION(3, 0, 99) <= ZEPHYR_VERSION_CODE
-	/* During kernel initialization, the next pm state is set to ACTIVE. By checking this
+	/* Zephyr implementations on secondary cores initialzation are
+	 * different between MTL+ and other platforms.
+	 */
+#if defined(CONFIG_ACE)
+	/* On MTL+ platforms, only run z_init_cpu() on boot, not when waking up from D3,
+	 * regardless whether saving to IMR is enabled or not.
```
Comment (lyakh) [inline zephyr/lib/cpu.c:151]: this is not how I understood the flow in a discussion yesterday. If on MTL+ IMR context saving isn't enabled, then waking up from D3 will be equivalent to the first boot, so the `pm_state_next_get(id)->state == PM_STATE_ACTIVE` condition will be true and `z_init_cpu()` will be run.

**Review by lyakh — CHANGES_REQUESTED** [review]

File: zephyr/lib/cpu.c (hunk 2 above)
Comment (mengdonglin) [reply zephyr/lib/cpu.c]: > this is not how I understood the flow in a discussion yesterday. If on MTL+ IMR context saving isn't enabled, then waking up from D3 will be equivalent to the first boot, so the `pm_state_next_get(id)->state == PM_STATE_ACTIVE` condition will be true and `z_init_cpu()` will be run.

@tmleman Can you help clarify?
I also updated the PR according to your https://github.com/thesofproject/sof/commit/fa1572c4d5668bda0867d46c687808bcdd800458 for mtl-006.

Comment (tmleman) [reply zephyr/lib/cpu.c]: On MTL `z_init_cpu(id)` is called for secondary cores only for the first init. As I wrote in the comment in the code:

> During kernel initialization, the next pm state is set to ACTIVE.

If core goes from D0 to D3, next state is set to `PM_STATE_SOFT_OFF`. Thanks to this we know if the core is enabled for the first time.

`CONFIG_ADSP_IMR_CONTEXT_SAVE` changes the behavior of the primary core during D3. If disabled, every exit from D3 for primary core is like first boot. Therefore, if we re-enable the secondary core after that, the function `z_init_cpu(id)` will be called.

File: zephyr/lib/cpu.c [hunk 3]
```diff
@@ -143,12 +143,19 @@ int cpu_enable_core(int id)
 		return 0;
 
 #if ZEPHYR_VERSION(3, 0, 99) <= ZEPHYR_VERSION_CODE
-	/* During kernel initialization, the next pm state is set to ACTIVE. By checking this
+	/* Zephyr implementations on secondary cores initialzation are
+	 * different between MTL+ and other platforms.
+	 *
+	 * On MTL+ platforms, only run z_init_cpu() on boot, not when waking up from D3,
+	 * regardless whether saving to IMR is enabled or not.
```
Comment (plbossart) [inline zephyr/lib/cpu.c:150]: that seems wrong to me @mengdonglin 

What is the difference between 'boot'  and 'waking up from D3' if there is no context saved in IMR?

**Review by plbossart — CHANGES_REQUESTED** [review]
I don't think the comments are quite right.

**Comment by kv2019i** [comment]
So seems we are only discussng about the comments now. We do really need to get this in to keep to the SOF2.7 timeline. 

Let me propose alternative wording based on comments so far.
https://github.com/thesofproject/sof/pull/8170

**Comment by marc-hb** [comment]
> So seems we are only discussing about the comments now.

Ignoring... test failures?
- https://github.com/thesofproject/sof/pull/8170#issuecomment-1709010997

**Comment by kv2019i** [comment]
Merged via https://github.com/thesofproject/sof/pull/8170 

We'll follow-up to clarify the z_init_cpu() code (and comments) in mainline to get rid of platform specific checks on SOF side.

**Pull request closed without merging**